File: Frasy/lua/core/utils/is.lua
--- @file    is.lua
--- @author  Paul Thomas
--- @date    2026-06-03
---
--- @copyright
--- This program is free software: you can redistribute it and/or modify it under the
--- terms of the GNU General Public License as published by the Free Software Foundation, either
--- version 3 of the License, or (at your option) any later version.
--- This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
--- even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
--- General Public License for more details.
--- You should have received a copy of the GNU General Public License along with this program. If
--- not, see <a href=https://www.gnu.org/licenses/>https://www.gnu.org/licenses/</a>.

local Is = {}

local function isOdIn(od) return Is.Table(od) and Is.Not(od.lowLimit, Is.Nil) and Is.Not(od.highLimit, Is.Nil) end

function Is.Array(v) return Is.Table(v) and #v ~= 0 end

function Is.ArrayInOd(v, od) return Is.Array(v) and #v <= od.stringLengthMin end

function Is.InArray(v, t)
    if not Is.Array(v) then return false end
    for _, e in ipairs(t) do
        if type(e) == type(v) and e == v then return true end
    end
    return false
end

function Is.Boolean(v) return type(v) == "boolean" end

function Is.Float(v) return type(v) == "number" end

function Is.FloatIn(v, min, max) return Is.Float(v) and min <= v and v <= max end

function Is.FloatInEx(v, min, max) return Is.Float(v) and min < v and v < max end

function Is.FloatInOd(v, od) return isOdIn(od) and Is.FloatIn(v, od.lowLimit, od.highLimit) end

function Is.Function(v) return type(v) == "function" end

function Is.Integer(v) return Is.Number(v) and (v // 1) == v end

function Is.Integer8(v) return Is.IntegerIn(v, -128, 127) end

function Is.Integer16(v) return Is.IntegerIn(v, -32768, 32767) end

function Is.Integer32(v) return Is.IntegerIn(v, -2147483648, 2147483647) end

function Is.IntegerIn(v, min, max) return Is.Integer(v) and min <= v and v <= max end

function Is.IntegerInEx(v, min, max) return Is.Integer(v) and min < v and v < max end

function Is.IntegerInOd(v, od) return isOdIn(od) and Is.IntegerIn(v, od.lowLimit, od.highLimit) end

function Is.Nil(v) return v == nil end

function Is.Not(v, f, ...) return not f(v, ...) end

function Is.Number(v) return type(v) == "number" end

function Is.String(v) return type(v) == "string" end

function Is.Table(v) return type(v) == "table" end

function Is.Unsigned(v) return Is.Integer(v) and 0 <= v end

function Is.Unsigned8(v) return Is.UnsignedIn(v, 0, 255) end

function Is.Unsigned16(v) return Is.UnsignedIn(v, 0, 65535) end

function Is.Unsigned32(v) return Is.UnsignedIn(v, 0, 4294967295) end

function Is.UnsignedIn(v, min, max) return Is.Unsigned(v) and min <= v and v <= max end

function Is.UnsignedInEx(v, min, max) return Is.Unsigned(v) and min < v and v < max end

function Is.UnsignedInOd(v, od) return isOdIn(od) and Is.UnsignedIn(v, od.lowLimit, od.highLimit) end

return Is

File: tests/utils/is.cpp
#include "lua_test_fixture.h"

#include <gtest/gtest.h>

// =============================================================================
// Is module
// =============================================================================

class IsTest : public LuaTestFixture
{
protected:
    void SetUp() override
    {
        LuaTestFixture::SetUp();
        lua.script("Is = require('lua/core/utils/is')");
    }
};

TEST_F(IsTest, TypeChecks)
{
    EXPECT_TRUE(lua.script("return Is.Integer(5)").get<bool>());
    EXPECT_TRUE(lua.script("return Is.Integer(-10)").get<bool>());
    EXPECT_FALSE(lua.script("return Is.Integer(1.5)").get<bool>());
    EXPECT_FALSE(lua.script("return Is.Integer('5')").get<bool>());

    EXPECT_TRUE(lua.script("return Is.Float(1.5)").get<bool>());
    EXPECT_TRUE(lua.script("return Is.Float(1)").get<bool>()); // integers are numbers in Lua
    EXPECT_FALSE(lua.script("return Is.Float('1.5')").get<bool>());

    EXPECT_TRUE(lua.script("return Is.String('hello')").get<bool>());
    EXPECT_FALSE(lua.script("return Is.String(123)").get<bool>());

    EXPECT_TRUE(lua.script("return Is.Boolean(true)").get<bool>());
    EXPECT_TRUE(lua.script("return Is.Boolean(false)").get<bool>());
    EXPECT_FALSE(lua.script("return Is.Boolean(1)").get<bool>());

    EXPECT_TRUE(lua.script("return Is.Nil(nil)").get<bool>());
    EXPECT_FALSE(lua.script("return Is.Nil(0)").get<bool>());

    EXPECT_TRUE(lua.script("return Is.Table({})").get<bool>());
    EXPECT_TRUE(lua.script("return Is.Table({1,2,3})").get<bool>());
    EXPECT_FALSE(lua.script("return Is.Table('not a table')").get<bool>());

    EXPECT_TRUE(lua.script("return Is.Function(print)").get<bool>());
    EXPECT_FALSE(lua.script("return Is.Function(42)").get<bool>());
}

TEST_F(IsTest, Array)
{
    EXPECT_TRUE(lua.script("return Is.Array({1, 2, 3})").get<bool>());
    EXPECT_FALSE(lua.script("return Is.Array({})").get<bool>()); // empty table has #t == 0
    EXPECT_FALSE(lua.script("return Is.Array('hello')").get<bool>());
}

TEST_F(IsTest, Number)
{
    EXPECT_TRUE(lua.script("return Is.Number(42)").get<bool>());
    EXPECT_TRUE(lua.script("return Is.Number(3.14)").get<bool>());
    EXPECT_FALSE(lua.script("return Is.Number('42')").get<bool>());
}

TEST_F(IsTest, Unsigned)
{
    EXPECT_TRUE(lua.script("return Is.Unsigned(0)").get<bool>());
    EXPECT_TRUE(lua.script("return Is.Unsigned(100)").get<bool>());
    EXPECT_FALSE(lua.script("return Is.Unsigned(-1)").get<bool>());
    EXPECT_FALSE(lua.script("return Is.Unsigned(1.5)").get<bool>());
}

TEST_F(IsTest, Integer8)
{
    EXPECT_TRUE(lua.script("return Is.Integer8(-128)").get<bool>());
    EXPECT_TRUE(lua.script("return Is.Integer8(127)").get<bool>());
    EXPECT_TRUE(lua.script("return Is.Integer8(0)").get<bool>());
    EXPECT_FALSE(lua.script("return Is.Integer8(-129)").get<bool>());
    EXPECT_FALSE(lua.script("return Is.Integer8(128)").get<bool>());
}

TEST_F(IsTest, Integer16)
{
    EXPECT_TRUE(lua.script("return Is.Integer16(-32768)").get<bool>());
    EXPECT_TRUE(lua.script("return Is.Integer16(32767)").get<bool>());
    EXPECT_FALSE(lua.script("return Is.Integer16(-32769)").get<bool>());
    EXPECT_FALSE(lua.script("return Is.Integer16(32768)").get<bool>());
}

TEST_F(IsTest, Integer32)
{
    EXPECT_TRUE(lua.script("return Is.Integer32(-2147483648)").get<bool>());
    EXPECT_TRUE(lua.script("return Is.Integer32(2147483647)").get<bool>());
    EXPECT_FALSE(lua.script("return Is.Integer32(-2147483649)").get<bool>());
    EXPECT_FALSE(lua.script("return Is.Integer32(2147483648)").get<bool>());
}

TEST_F(IsTest, Unsigned8)
{
    EXPECT_TRUE(lua.script("return Is.Unsigned8(0)").get<bool>());
    EXPECT_TRUE(lua.script("return Is.Unsigned8(255)").get<bool>());
    EXPECT_FALSE(lua.script("return Is.Unsigned8(-1)").get<bool>());
    EXPECT_FALSE(lua.script("return Is.Unsigned8(256)").get<bool>());
}

TEST_F(IsTest, Unsigned16)
{
    EXPECT_TRUE(lua.script("return Is.Unsigned16(0)").get<bool>());
    EXPECT_TRUE(lua.script("return Is.Unsigned16(65535)").get<bool>());
    EXPECT_FALSE(lua.script("return Is.Unsigned16(-1)").get<bool>());
    EXPECT_FALSE(lua.script("return Is.Unsigned16(65536)").get<bool>());
}

TEST_F(IsTest, Unsigned32)
{
    EXPECT_TRUE(lua.script("return Is.Unsigned32(0)").get<bool>());
    EXPECT_TRUE(lua.script("return Is.Unsigned32(4294967295)").get<bool>());
    EXPECT_FALSE(lua.script("return Is.Unsigned32(-1)").get<bool>());
    EXPECT_FALSE(lua.script("return Is.Unsigned32(4294967296)").get<bool>());
}

TEST_F(IsTest, IntegerIn)
{
    EXPECT_TRUE(lua.script("return Is.IntegerIn(5, 0, 10)").get<bool>());
    EXPECT_TRUE(lua.script("return Is.IntegerIn(0, 0, 10)").get<bool>());  // boundary
    EXPECT_TRUE(lua.script("return Is.IntegerIn(10, 0, 10)").get<bool>()); // boundary
    EXPECT_FALSE(lua.script("return Is.IntegerIn(-1, 0, 10)").get<bool>());
    EXPECT_FALSE(lua.script("return Is.IntegerIn(11, 0, 10)").get<bool>());
    EXPECT_FALSE(lua.script("return Is.IntegerIn(1.5, 0, 10)").get<bool>()); // not integer
}

TEST_F(IsTest, IntegerInEx)
{
    EXPECT_TRUE(lua.script("return Is.IntegerInEx(5, 0, 10)").get<bool>());
    EXPECT_FALSE(lua.script("return Is.IntegerInEx(0, 0, 10)").get<bool>());  // exclusive
    EXPECT_FALSE(lua.script("return Is.IntegerInEx(10, 0, 10)").get<bool>()); // exclusive
}

TEST_F(IsTest, FloatIn)
{
    EXPECT_TRUE(lua.script("return Is.FloatIn(5.5, 0, 10)").get<bool>());
    EXPECT_TRUE(lua.script("return Is.FloatIn(0, 0, 10)").get<bool>());   // boundary
    EXPECT_TRUE(lua.script("return Is.FloatIn(10, 0, 10)").get<bool>());  // boundary
    EXPECT_FALSE(lua.script("return Is.FloatIn(-0.1, 0, 10)").get<bool>());
    EXPECT_FALSE(lua.script("return Is.FloatIn(10.1, 0, 10)").get<bool>());
    EXPECT_FALSE(lua.script("return Is.FloatIn('5', 0, 10)").get<bool>()); // not a number
}

TEST_F(IsTest, FloatInEx)
{
    EXPECT_TRUE(lua.script("return Is.FloatInEx(5.5, 0, 10)").get<bool>());
    EXPECT_FALSE(lua.script("return Is.FloatInEx(0, 0, 10)").get<bool>());  // exclusive
    EXPECT_FALSE(lua.script("return Is.FloatInEx(10, 0, 10)").get<bool>()); // exclusive
}

TEST_F(IsTest, UnsignedIn)
{
    EXPECT_TRUE(lua.script("return Is.UnsignedIn(5, 0, 10)").get<bool>());
    EXPECT_TRUE(lua.script("return Is.UnsignedIn(0, 0, 10)").get<bool>());
    EXPECT_TRUE(lua.script("return Is.UnsignedIn(10, 0, 10)").get<bool>());
    EXPECT_FALSE(lua.script("return Is.UnsignedIn(11, 0, 10)").get<bool>());
    EXPECT_FALSE(lua.script("return Is.UnsignedIn(-1, 0, 10)").get<bool>());
}

TEST_F(IsTest, Not)
{
    EXPECT_TRUE(lua.script("return Is.Not(nil, Is.Integer)").get<bool>());
    EXPECT_TRUE(lua.script("return Is.Not('hello', Is.Integer)").get<bool>());
    EXPECT_FALSE(lua.script("return Is.Not(5, Is.Integer)").get<bool>());
}

TEST_F(IsTest, InArray)
{
    EXPECT_TRUE(lua.script("return Is.InArray(2, {1, 2, 3})").get<bool>());
    EXPECT_FALSE(lua.script("return Is.InArray(4, {1, 2, 3})").get<bool>());
    EXPECT_FALSE(lua.script("return Is.InArray(2, {})").get<bool>()); // empty array fails Is.Array
    EXPECT_TRUE(lua.script("return Is.InArray('b', {'a', 'b', 'c'})").get<bool>());
    EXPECT_FALSE(lua.script("return Is.InArray('d', {'a', 'b', 'c'})").get<bool>());
}
